File: generate_cover_sheet.js
const {
  Document, Packer, Paragraph, TextRun, AlignmentType,
  BorderStyle, Table, TableRow, TableCell, WidthType,
  ShadingType, VerticalAlign, TabStopType
} = require('docx');
const fs = require('fs');

const data = JSON.parse(process.argv[2]);
const outputPath = process.argv[3];

const m = data.merchant;
const b = data.bank;
const today = data.date || new Date().toLocaleDateString('en-US', { month: '2-digit', day: '2-digit', year: 'numeric' });
const companyName = m.company_name;
const dba = m.dba_name && m.dba_name !== m.company_name ? m.dba_name : null;
const ownerName = m.entity_creator_name;
const state = m.company_state;
const address = `${m.company_address}, ${m.company_city}, ${m.company_state} ${m.company_zip}`;
const phone = m.location_phone || '';
const email = m.email || '';
const fein = m.fein || '';

// ── Helpers ────────────────────────────────────────────────────────────────

const BLUE = '1A3C5E';
const LIGHT_BLUE = 'EAF1F8';
const MID_BLUE = '2C5F8A';
const GRAY = '666666';
const RULE_COLOR = 'BBCFE0';

function spacer(after = 80) {
  return new Paragraph({ spacing: { before: 0, after }, children: [] });
}

const cellBorder = { style: BorderStyle.SINGLE, size: 4, color: RULE_COLOR };
const borders = { top: cellBorder, bottom: cellBorder, left: cellBorder, right: cellBorder };
const noBorder = { style: BorderStyle.NONE, size: 0, color: 'FFFFFF' };
const noBorders = { top: noBorder, bottom: noBorder, left: noBorder, right: noBorder };
const cellPad = { top: 60, bottom: 60, left: 140, right: 140 };

function infoRow(label, value) {
  return new TableRow({
    children: [
      new TableCell({
        borders,
        margins: cellPad,
        width: { size: 1600, type: WidthType.DXA },
        shading: { fill: LIGHT_BLUE, type: ShadingType.CLEAR },
        children: [new Paragraph({
          children: [new TextRun({ text: label, font: 'Arial', size: 15, bold: true, color: BLUE })]
        })]
      }),
      new TableCell({
        borders,
        margins: cellPad,
        width: { size: 6960, type: WidthType.DXA },
        children: [new Paragraph({
          children: [new TextRun({ text: value || '—', font: 'Arial', size: 15, color: value ? '000000' : GRAY })]
        })]
      }),
    ]
  });
}

function stepRow(num, title, items, statusLabel = '□ Pending') {
  const statusColor = statusLabel.includes('✓') ? '1A6B3A' : statusLabel.includes('●') ? '8B5E00' : GRAY;
  return new TableRow({
    children: [
      // Step number
      new TableCell({
        borders,
        margins: { top: 120, bottom: 120, left: 140, right: 140 },
        width: { size: 600, type: WidthType.DXA },
        shading: { fill: BLUE, type: ShadingType.CLEAR },
        verticalAlign: VerticalAlign.TOP,
        children: [new Paragraph({
          alignment: AlignmentType.CENTER,
          children: [new TextRun({ text: String(num), font: 'Arial', size: 26, bold: true, color: 'FFFFFF' })]
        })]
      }),
      // Title + items
      new TableCell({
        borders,
        margins: { top: 100, bottom: 100, left: 160, right: 160 },
        width: { size: 6860, type: WidthType.DXA },
        verticalAlign: VerticalAlign.TOP,
        children: [
          new Paragraph({
            spacing: { before: 0, after: 60 },
            children: [new TextRun({ text: title, font: 'Arial', size: 16, bold: true, color: BLUE })]
          }),
          ...items.map(item => new Paragraph({
            spacing: { before: 0, after: 20 },
            children: [new TextRun({ text: `  ${item}`, font: 'Arial', size: 15, color: '333333' })]
          }))
        ]
      }),
      // Status + date
      new TableCell({
        borders,
        margins: { top: 100, bottom: 100, left: 140, right: 140 },
        width: { size: 1500, type: WidthType.DXA },
        shading: { fill: 'F7FAFD', type: ShadingType.CLEAR },
        verticalAlign: VerticalAlign.TOP,
        children: [
          new Paragraph({
            alignment: AlignmentType.CENTER,
            spacing: { before: 0, after: 80 },
            children: [new TextRun({ text: statusLabel, font: 'Arial', size: 15, bold: true, color: statusColor })]
          }),
          new Paragraph({
            alignment: AlignmentType.CENTER,
            children: [new TextRun({ text: 'Date: __________', font: 'Arial', size: 15, color: GRAY })]
          }),
        ]
      }),
    ]
  });
}

function docCheckRow(docName, included = true) {
  return new TableRow({
    children: [
      new TableCell({
        borders,
        margins: cellPad,
        width: { size: 600, type: WidthType.DXA },
        shading: { fill: included ? 'EAF7EF' : 'FEF9E7', type: ShadingType.CLEAR },
        children: [new Paragraph({
          alignment: AlignmentType.CENTER,
          children: [new TextRun({ text: included ? '☑' : '☐', font: 'Arial', size: 16, color: included ? '1A6B3A' : '8B5E00' })]
        })]
      }),
      new TableCell({
        borders,
        margins: cellPad,
        width: { size: 8760, type: WidthType.DXA },
        children: [new Paragraph({
          children: [new TextRun({ text: docName, font: 'Arial', size: 16, color: '222222' })]
        })]
      }),
    ]
  });
}

// ── Document ───────────────────────────────────────────────────────────────

const doc = new Document({
  styles: { default: { document: { run: { font: 'Arial', size: 16 } } } },
  sections: [{
    properties: {
      page: {
        size: { width: 12240, height: 15840 },
        margin: { top: 720, right: 900, bottom: 720, left: 900 }
      }
    },
    children: [

      // ── Header bar (simulated with top border paragraph) ───────────────
      new Paragraph({
        spacing: { before: 0, after: 0 },
        border: { top: { style: BorderStyle.SINGLE, size: 32, color: BLUE, space: 0 } },
        shading: { fill: BLUE, type: ShadingType.CLEAR },
        children: [new TextRun({ text: ' ', font: 'Arial', size: 4 })]
      }),
      new Paragraph({
        spacing: { before: 0, after: 0 },
        shading: { fill: BLUE, type: ShadingType.CLEAR },
        children: [new TextRun({ text: '  ATM MERCHANT ONBOARDING', font: 'Arial', size: 16, bold: true, color: 'FFFFFF' })]
      }),
      new Paragraph({
        spacing: { before: 0, after: 0 },
        shading: { fill: BLUE, type: ShadingType.CLEAR },
        children: [new TextRun({ text: '  PACKET COVER SHEET', font: 'Arial', size: 16, bold: true, color: 'FFFFFF' })]
      }),
      new Paragraph({
        spacing: { before: 0, after: 200 },
        shading: { fill: BLUE, type: ShadingType.CLEAR },
        border: { bottom: { style: BorderStyle.SINGLE, size: 32, color: BLUE, space: 0 } },
        children: [new TextRun({ text: ' ', font: 'Arial', size: 4 })]
      }),

      // ── Merchant Info ──────────────────────────────────────────────────
      new Paragraph({
        spacing: { before: 0, after: 100 },
        children: [new TextRun({ text: 'MERCHANT INFORMATION', font: 'Arial', size: 16, bold: true, color: MID_BLUE })]
      }),

      new Table({
        width: { size: 9360, type: WidthType.DXA },
        columnWidths: [2400, 6960],
        rows: [
          infoRow('Legal Entity Name', companyName),

          infoRow('Owner / Managing Member', ownerName),
          infoRow('State of Formation', state),
          infoRow('Principal Address', address),
          infoRow('Phone', phone),
          infoRow('Email', email),
          infoRow('FEIN', fein || 'Pending — see EIN Answer Sheet'),
          infoRow('Bank', b.bank_name || ''),
          infoRow('Date Prepared', today),
        ]
      }),

      spacer(80),

      // ── 4-Step Status Tracker ──────────────────────────────────────────
      new Paragraph({
        spacing: { before: 0, after: 100 },
        children: [new TextRun({ text: 'ONBOARDING STATUS', font: 'Arial', size: 16, bold: true, color: MID_BLUE })]
      }),

      new Table({
        width: { size: 9360, type: WidthType.DXA },
        columnWidths: [600, 6860, 1900],
        rows: [
          stepRow(1, 'File for LLC', [
            '□ Articles of Organization submitted',
            '□ State filing fee paid',
            '□ Approval documents received',
            '□ Operating Agreement drafted',
          ]),
          stepRow(2, 'Obtain EIN', [
            '□ IRS online application completed (irs.gov/ein)',
            '□ CP 575 confirmation letter downloaded',
            '□ EIN recorded: ____________________________',
          ]),
          stepRow(3, 'Open Bank Account', [
            '□ Business checking account opened',
            '□ Routing number obtained',
            '□ Account number obtained',
            '□ Banking Relationship Letter signed & stamped',
          ]),
          stepRow(4, 'Submit Processing Forms', [
            '□ Exhibit 2 — ATM Operator Agreement signed',
            '□ Exhibit 3 — ACH Authorization Form signed',
            '□ W-9 signed',
            '□ Full packet submitted to ISO / PAI',
          ]),
        ]
      }),

      // ── Documents in Packet ────────────────────────────────────────────
      new Paragraph({
        spacing: { before: 0, after: 100 },
        children: [new TextRun({ text: 'DOCUMENTS IN THIS PACKET', font: 'Arial', size: 16, bold: true, color: MID_BLUE })]
      }),

      new Table({
        width: { size: 9360, type: WidthType.DXA },
        columnWidths: [600, 8760],
        rows: [
          docCheckRow('Exhibit 2 — ATM Operator / Source of Funds Provider Agreement'),
          docCheckRow('Exhibit 3 — ACH Authorization Form'),
          docCheckRow('W-9 — Taxpayer Identification'),
          docCheckRow('Banking Relationship Letter'),
          docCheckRow('EIN Application Answer Sheet'),
        ]
      }),

    ]
  }]
});

Packer.toBuffer(doc).then(buffer => {
  fs.writeFileSync(outputPath, buffer);
  console.log('OK: ' + outputPath);
}).catch(err => {
  console.error('ERROR: ' + err.message);
  process.exit(1);
});

File: generate_cover_sheet.py
"""Cover Sheet generator — pure Python (replaces generate_cover_sheet.js)"""
import sys, json
from docx import Document
from docx.shared import Pt, Inches, RGBColor
from docx.enum.text import WD_ALIGN_PARAGRAPH
from docx.oxml.ns import qn
from docx.oxml import OxmlElement


BLUE     = '1A3C5E'
MID_BLUE = '2C5F8A'
LT_BLUE  = 'EAF1F8'
GRAY     = '666666'
RULE     = 'BBCFE0'
GREEN    = '1A6B3A'
LT_GREEN = 'EAF7EF'


def rgb(h):
    return RGBColor(int(h[0:2],16), int(h[2:4],16), int(h[4:6],16))


def set_cell_bg(cell, hex_color):
    tc = cell._tc
    tcPr = tc.get_or_add_tcPr()
    shd = OxmlElement('w:shd')
    shd.set(qn('w:fill'), hex_color)
    shd.set(qn('w:val'), 'clear')
    tcPr.append(shd)


def cell_borders(cell, color=RULE):
    tc = cell._tc
    tcPr = tc.get_or_add_tcPr()
    tcBorders = OxmlElement('w:tcBorders')
    for side in ['top','left','bottom','right']:
        el = OxmlElement(f'w:{side}')
        el.set(qn('w:val'), 'single')
        el.set(qn('w:sz'), '4')
        el.set(qn('w:color'), color)
        tcBorders.append(el)
    tcPr.append(tcBorders)


def add_run(para, text, bold=False, size=8, color=None):
    run = para.add_run(text)
    run.bold = bold
    run.font.name = 'Arial'
    run.font.size = Pt(size)
    if color:
        run.font.color.rgb = rgb(color)
    return run


def section_heading(doc, text):
    p = doc.add_paragraph()
    p.paragraph_format.space_before = Pt(8)
    p.paragraph_format.space_after = Pt(4)
    add_run(p, text, bold=True, size=8, color=MID_BLUE)


def info_table(doc, rows):
    table = doc.add_table(rows=0, cols=2)
    table.style = 'Table Grid'
    for label, value in rows:
        row = table.add_row()
        c0, c1 = row.cells[0], row.cells[1]
        set_cell_bg(c0, LT_BLUE)
        cell_borders(c0)
        cell_borders(c1)
        p0 = c0.paragraphs[0]
        p0.paragraph_format.space_after = Pt(0)
        add_run(p0, label, bold=True, size=8, color=BLUE)
        p1 = c1.paragraphs[0]
        p1.paragraph_format.space_after = Pt(0)
        add_run(p1, value or '—', size=8, color='000000' if value else GRAY)
    doc.add_paragraph().paragraph_format.space_after = Pt(4)


def step_table(doc, steps):
    """steps: list of (num, title, items)"""
    table = doc.add_table(rows=0, cols=3)
    table.style = 'Table Grid'
    for num, title, items in steps:
        row = table.add_row()
        c0, c1, c2 = row.cells[0], row.cells[1], row.cells[2]

        # Number cell
        set_cell_bg(c0, BLUE)
        cell_borders(c0, BLUE)
        p0 = c0.paragraphs[0]
        p0.alignment = WD_ALIGN_PARAGRAPH.CENTER
        p0.paragraph_format.space_after = Pt(0)
        add_run(p0, str(num), bold=True, size=13, color='FFFFFF')

        # Content cell
        cell_borders(c1)
        p_title = c1.paragraphs[0]
        p_title.paragraph_format.space_after = Pt(2)
        add_run(p_title, title, bold=True, size=8, color=BLUE)
        for item in items:
            p_item = c1.add_paragraph()
            p_item.paragraph_format.space_after = Pt(1)
            add_run(p_item, f'  {item}', size=8, color='333333')

        # Status cell
        set_cell_bg(c2, 'F7FAFD')
        cell_borders(c2)
        p_status = c2.paragraphs[0]
        p_status.alignment = WD_ALIGN_PARAGRAPH.CENTER
        p_status.paragraph_format.space_after = Pt(4)
        add_run(p_status, '□ Pending', bold=True, size=8, color=GRAY)
        p_date = c2.add_paragraph()
        p_date.alignment = WD_ALIGN_PARAGRAPH.CENTER
        p_date.paragraph_format.space_after = Pt(0)
        add_run(p_date, 'Date: __________', size=8, color=GRAY)

    doc.add_paragraph().paragraph_format.space_after = Pt(4)


def checklist_table(doc, docs):
    table = doc.add_table(rows=0, cols=2)
    table.style = 'Table Grid'
    for name in docs:
        row = table.add_row()
        c0, c1 = row.cells[0], row.cells[1]
        set_cell_bg(c0, LT_GREEN)
        cell_borders(c0)
        cell_borders(c1)
        p0 = c0.paragraphs[0]
        p0.alignment = WD_ALIGN_PARAGRAPH.CENTER
        p0.paragraph_format.space_after = Pt(0)
        add_run(p0, '☑', bold=True, size=8, color=GREEN)
        p1 = c1.paragraphs[0]
        p1.paragraph_format.space_after = Pt(0)
        add_run(p1, name, size=8)


def generate(data, output_path):
    m = data['merchant']
    b = data.get('bank', {})
    today = data.get('date', '')
    company_name = m['company_name']
    owner_name = m.get('entity_creator_name', '')
    state = m.get('company_state', '')
    address = f"{m.get('company_address','')}, {m.get('company_city','')}, {state} {m.get('company_zip','')}"
    phone = m.get('location_phone', '')
    email = m.get('email', '')


    doc = Document()
    for section in doc.sections:
        section.top_margin = Inches(0.5)
        section.bottom_margin = Inches(0.5)
        section.left_margin = Inches(0.625)
        section.right_margin = Inches(0.625)

    # Header bar
    hdr = doc.add_paragraph()
    hdr.paragraph_format.space_before = Pt(0)
    hdr.paragraph_format.space_after = Pt(2)
    pPr = hdr._p.get_or_add_pPr()
    shd = OxmlElement('w:shd')
    shd.set(qn('w:fill'), BLUE)
    shd.set(qn('w:val'), 'clear')
    pPr.append(shd)
    add_run(hdr, '  ATM MERCHANT ONBOARDING  |  PACKET COVER SHEET', bold=True, size=12, color='FFFFFF')

    doc.add_paragraph().paragraph_format.space_after = Pt(4)

    section_heading(doc, 'MERCHANT INFORMATION')
    info_table(doc, [
        ('Legal Entity Name', company_name),
        ('Owner / Managing Member', owner_name),
        ('State of Formation', state),
        ('Principal Address', address),
        ('Phone', phone),
        ('Email', email),
        ('Date Prepared', today),
    ])

    section_heading(doc, 'ONBOARDING STATUS')
    step_table(doc, [
        (1, 'File for LLC', [
            '□ Articles of Organization submitted',
            '□ State filing fee paid',
            '□ Approval documents received',
            '□ Operating Agreement drafted',
        ]),
        (2, 'Obtain EIN', [
            '□ IRS online application completed (irs.gov/ein)',
            '□ CP 575 confirmation letter downloaded',
            '□ EIN recorded: ____________________________',
        ]),
        (3, 'Open Bank Account', [
            '□ Business checking account opened',
            '□ Routing number obtained',
            '□ Account number obtained',
            '□ Banking Relationship Letter signed by banker',
        ]),
        (4, 'Emailing the Bank Letter and EIN Letter', [
            '□ Email the Bank Letter',
            '□ Email the EIN Letter',
        ]),
    ])

    doc.save(output_path)
    print(f'OK: {output_path}')


if __name__ == '__main__':
    data = json.loads(sys.argv[1])
    output_path = sys.argv[2]
    generate(data, output_path)
